pkg/kgo/ring.go

```diff
@@ -101,169 +107,88 @@ func (r *ringReq) dropPeek() (next promisedReq, more, dead bool) {
-// because we do not use it, but we keep `c` for testing lowering eight/mask7.
-type ringSeqResp struct {
-	mu sync.Mutex
-	c  *sync.Cond
-
-	elems [eight]*seqResp
-
-	head uint8
-	tail uint8
-	l    uint8
-}
-
-func (r *ringSeqResp) push(sr *seqResp) (first bool) {
-	r.mu.Lock()
-	defer r.mu.Unlock()
-
-	for r.l == eight {
-		if r.c == nil {
-			r.c = sync.NewCond(&r.mu)
-		}
-		r.c.Wait()
-	}
-
-	r.elems[r.tail] = sr
-	r.tail = (r.tail + 1) & mask7
-	r.l++
-
-	return r.l == 1
-}
-
-func (r *ringSeqResp) dropPeek() (next *seqResp, more bool) {
-	r.mu.Lock()
-	defer r.mu.Unlock()
-
-	r.elems[r.head] = nil
-	r.head = (r.head + 1) & mask7
-	r.l--
-
-	if r.c != nil {
-		r.c.Signal()
-	}
-
-	return r.elems[r.head], r.l > 0
-}
-
-// Also no die; this type is slightly different because we can have overflow.
-// If we have overflow, we add to overflow until overflow is drained -- we
-// always want strict odering.
-type ringBatchPromise struct {
-	mu sync.Mutex
-
-	elems [eight]batchPromise
-
-	head uint8
-	tail uint8
-	l    uint8
-
-	overflow []batchPromise
-}
-
-func (r *ringBatchPromise) push(b batchPromise) (first bool) {
-	r.mu.Lock()
-	defer r.mu.Unlock()
-
 	// If the ring is full, we go into overflow; if overflow is non-empty,
 	// for ordering purposes, we add to the end of overflow. We only go
 	// back to using the ring once overflow is finally empty.
 	if r.l == eight || len(r.overflow) > 0 {
-		r.overflow = append(r.overflow, b)
-		return false
+		r.overflow = append(r.overflow, elem)
+		return false, false
 	}
 
-	r.elems[r.tail] = b
+	r.elems[r.tail] = elem
 	r.tail = (r.tail + 1) & mask7
 	r.l++
 
-	return r.l == 1
+	return r.l == 1, false
 }
 
-func (r *ringBatchPromise) dropPeek() (next batchPromise, more bool) {
+func (r *unlimitedRing[T]) dropPeek() (next T, more, dead bool) {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 
 	// We always drain the ring first. If the ring is ever empty, there
 	// must be overflow: we would not be here if the ring is not-empty.
 	if r.l > 1 {
-		r.elems[r.head] = batchPromise{}
+		var zero T
+		r.elems[r.head] = zero
 		r.head = (r.head + 1) & mask7
 		r.l--
-		return r.elems[r.head], true
+		return r.elems[r.head], true, r.dead
 	} else if r.l == 1 {
-		r.elems[r.head] = batchPromise{}
+		var zero T
+		r.elems[r.head] = zero
 		r.head = (r.head + 1) & mask7
 		r.l--
 		if len(r.overflow) == 0 {
-			return next, false
+			return next, false, r.dead
 		}
-		return r.overflow[0], true
+		return r.overflow[0], true, r.dead
 	}
 	r.overflow = r.overflow[1:]
+
+	// If the overflow slice is 8x the fixed ring size, and the overflow length
+	// is less than 1/8 of the capacity, then we do shrink the overflow slice
+	// to 2x the currently required capacity. This prevents edge cases where
+	// the overflow slice could grow indefinitely.
+	if cap(r.overflow) > 64 && len(r.overflow) < cap(r.overflow)/8 {
+		shrunk := make([]T, len(r.overflow), len(r.overflow)*2)
+		copy(shrunk, r.overflow)
+		r.overflow = shrunk
+	}
```

Comment:
l178 shifts the start of overflow by one, which should eventually recycle the slice. Perhaps if the length is zero it can still be holding onto an old pointer, so, maybe it's worth nil'ing overflow if the len is 0 after shifting.

That said, I don't think this code will ever fire. The capacity is decremented by one every time we pull from overflow because we are shifting our starting position, not the end.

Reply:
> That said, I don't think this code will ever fire. The capacity is decremented by one every time we pull from overflow because we are shifting our starting position, not the end.

You're right.

> l178 shifts the start of overflow by one, which should eventually recycle the slice

I think the actual recycling happens when future `append()` will re-allocate the slice memory. See the comment I've added to https://github.com/twmb/franz-go/pull/1057/commits/45707365807fcba4d74ca2b931d6ae9cf3c63764, where I've also removed this manual trimming logic. Makes sense to you?

Reply:
Yeah I meant something along those lines but probably worded it poorly. Thanks!

Reply:
It still may be worth adding
```
} else {
    r.overflow = nil
}
```

so that the overflow field is not hanging onto any empty but non-nil slice behind the scenes, replacing line 124 which just closes the conditional at the moment.